File: src/main/kotlin/com/stevenyin/actioninkotlin/chapter04/_04_10/Button.java
package com.stevenyin.actioninkotlin.chapter04._04_10;

import com.stevenyin.actioninkotlin.chapter04._04_09.State;
import com.stevenyin.actioninkotlin.chapter04._04_09.View;

public class Button implements View{

	public State getCurrentState() {
		return new ButtonState();
	}

	public void restoreState(State arg0) {
		
	}
	
	public class ButtonState implements State{}
	// ButtonState是Button的内部类, ButtonState隐式地存储了它的外部Button类的引用,
	// 当我们想要对ButtonState进行序列化的时候, 会出现java.io.NotSerializableException: Button
	// 由于Button是不可以序列化的, 所以破坏了ButtonState的序列化
	// 为了修复这个问题, 需要把ButtonState类声明成static的, 因为将一个内部类声明成static的, 会从这个类中删除掉他的Outer引用
}

File: src/main/kotlin/com/stevenyin/actioninkotlin/chapter04/_04_11/Button.kt
package com.stevenyin.actioninkotlin.chapter04._04_11

import com.stevenyin.actioninkotlin.chapter04._04_09.View
import com.stevenyin.actioninkotlin.chapter04._04_09.State
class Button: View {
	override fun getCurrentState() = ButtonState()
	override fun restoreState(state: State) {
		TODO()
	}
	class ButtonState: State {
	}
	// 对于Kotlin这种语法, 和4.10中添加static修饰后的语法意义是一样的(这里还相当于多了个final),
	// 要把它变成一个内部类来持有外部类的引用的话, 需要加上一个inner修饰符, 这样的话, 和4.10中不写static是一样的了
}